### PR TITLE
Setup docker buildx for all-in-one

### PR DESCRIPTION
**Pull request by morlay**
## Which problem is this PR solving?
- buildx in travis.yml to support #2292 

## Short description of the changes

```bash
# file ./scripts/travis/build-all-in-one-image.sh

bash ./scripts/travis/setup-docker-buildx.sh


# multi binary build
make build-all-in-one GOOS=linux GOARCH=amd64
make build-all-in-one GOOS=linux GOARCH=arm64

repo=jaegertracing/all-in-one

# load to local docker for integration_test
docker buildx build --load -f cmd/all-in-one/Dockerfile -t $repo:latest cmd/all-in-one
run_integration_test $repo
# push after testing
# could push multi tags by multi `-t $repo:latest -t $repo:1.18 -t $repo:1.18.1`
# it will load from buildkit cache
docker buildx build --push -f cmd/all-in-one/Dockerfile -t $repo:latest cmd/all-in-one
```

@pavolloffay 
have to change upload-to-docker, need to discuss.

### CODE REVIEW

**Comment by pavolloffay** [comment]
Does this build images for all architectures `docker buildx build --load -f cmd/all-in-one/Dockerfile -t $repo:latest cmd/all-in-one`?

**Comment by pavolloffay** [comment]
Could you please enable the multi arch build on travis and if possible reuse the existing make targets? Ideally we should not require dependency on buildx when building a single arch locally.

**Comment by morlay** [comment]
`docker buildx build --load -f --platform=linux/arm64,linux/amd64 cmd/all-in-one/Dockerfile -t $repo:latest cmd/all-in-one` will build arm64 & amd64 both.

but now in travis, need to run tests after building. this why `--load` here.

after testings, we could use `--push` to push mutli arch images.

and now, in your scripts, will tag for mutil versions.
and buildx not reuse the `./scripts/travis/upload-to-docker.sh`, i am no ideas should i refactor that file? or just refactor `./scripts/travis/build-all-in-one-image.sh`

**Comment by morlay** [comment]
@pavolloffay i updated `./scripts/travis/build-all-in-one-image.sh` with buildx.

**Comment by pavolloffay** [comment]
@morlay could you please open a separate PR with change from trusty to bionic?

**Comment by morlay** [comment]
@pavolloffay sure

File: scripts/travis/build-all-in-one-image.sh
Comment (pavolloffay) [inline scripts/travis/build-all-in-one-image.sh:49]: why has been this copied from `upload-to-docker.sh`? Could we just keep it there?

Comment (morlay) [reply scripts/travis/build-all-in-one-image.sh:49]: with buildx, we could not use tag directly.

`docker tag` just tag local image for local arch, when we push that tags, it will break the oci image with multi arch.

**Comment by pavolloffay** [comment]
@jaegertracing/jaeger-maintainers any objections on merging this? We could the multiarch build for only the all-in-one initially.

File: scripts/travis/build-all-in-one-image.sh
Comment (yurishkuro) [inline scripts/travis/build-all-in-one-image.sh:82]: nit: these vars are Make vars, not env vars, they can go on the right side.

File: scripts/travis/build-all-in-one-image.sh
Comment (yurishkuro) [reply scripts/travis/build-all-in-one-image.sh:49]: ok, but  there's a lot being copied here, and would have to be repeated for other Docker images.
Can we not either parameterize `upload-to-docker.sh` or refactor to avoid repetition?

File: scripts/travis/build-all-in-one-image.sh
Comment (yurishkuro) [inline scripts/travis/build-all-in-one-image.sh:47]: both of these are `$1`?

File: scripts/travis/build-all-in-one-image.sh
Comment (morlay) [reply scripts/travis/build-all-in-one-image.sh]: yep. i reverted it

File: scripts/travis/build-all-in-one-image.sh
Comment (morlay) [reply scripts/travis/build-all-in-one-image.sh]: Sorry. CMD_ROOT should be `$2`. fixed it.

File: scripts/travis/build-all-in-one-image.sh
Comment (pavolloffay) [reply scripts/travis/build-all-in-one-image.sh:49]: I am fine to go with this and see how it works and refactor other images in a second PR

**Comment by morlay** [comment]
@pavolloffay 

anything need i to do next?

**Review by pavolloffay — DISMISSED** [review]
I will approve it and merge next week. 

thanks @morlay!

**Comment by davidsbond** [comment]
@pavolloffay is there any chance to get this merged in? I'd like to be able to use the `allinone` image on my raspberry pi swarm

**Comment by morlay** [comment]
@davidsbond 
you could use https://hub.docker.com/r/querycapistio/all-in-one/tags before official support.

source here. https://github.com/querycap/istio/tree/master/build/jaegertracing

**Comment by jpkrohling** [comment]
@Mergifyio rebase

**Review by jpkrohling — COMMENTED** [review]
This looks nice, but we do need this documented somewhere in case we need to build the images locally. Would be nice to get positive confirmation from other reviewers (@yurishkuro and @pavolloffay) that you ran this locally and are able to build those images. The instructions from the jaeger-operator repo can be used as base for this.

Please, wait to work on the other images, as there's another PR that will conflict with this: #2545.

**Comment by morlay** [comment]
@jpkrohling 

local build here could just run `./scripts/travis/build-all-in-one-image.sh`
a sub script to setup buildx included.

jaeger-operator use github workflow for CI , so could use with actions.

conflicts #2545 could merged that pr first. 
i can do rebase and fix conflicts.

**Comment by jpkrohling** [comment]
> local build here could just run ./scripts/travis/build-all-in-one-image.sh
> a sub script to setup buildx included.

True, but we still need it documented somewhere. Not everyone is using Ubuntu, and we've seen that there are things specific to more secured distributions like Fedora that need to be done.

> conflicts #2545 could merged that pr first.
> i can do rebase and fix conflicts.

Thanks!

**Comment by morlay** [comment]
@jpkrohling 

But i am not familiar with Fedora

is the docker offical troubleshooting cover all issues?
https://docs.docker.com/engine/install/linux-postinstall/#troubleshooting

here a checker too.
```
$ curl https://raw.githubusercontent.com/docker/docker/master/contrib/check-config.sh > check-config.sh
$ bash ./check-config.sh
```

**Comment by jpkrohling** [comment]
Having something just like what we have for the operator would be sufficient: 

https://github.com/jaegertracing/jaeger-operator/blob/master/CONTRIBUTING.md#building-oci-images-for-multiple-arch-linuxarm64-linuxamd64

**Comment by morlay** [comment]
documents added. 

but `scripts/travis/build-all-in-one-image.sh` under travis folder. should we need to run local?

File: CONTRIBUTING.md
Comment (jpkrohling) [inline CONTRIBUTING.md:174]: Can we just go without this? If this is part of the official instructions, then let's just point users there.

**Comment by jpkrohling** [comment]
> but scripts/travis/build-all-in-one-image.sh under travis folder. should we need to run local?

No, it's fine to keep it there.

File: CONTRIBUTING.md
Comment (morlay) [reply CONTRIBUTING.md]: sure.

**Comment by jpkrohling** [comment]
>  i can do rebase and fix conflicts.

#2545 has just been merged. Do you want to update this PR?

**Comment by morlay** [comment]
Sure. i will updates this pr tomorrow.

**Comment by morlay** [comment]
@jpkrohling 

Have to remove `docker/*` and merge multi stages into `cmd/all-in-one/Dockerfile`
spliting base image builds have trouble for multi arch build (need to build base image for multi arch too.)

local run (after buildx ready):
```shell
TRAVIS_SECURE_ENV_VARS=true NAMESPACE=$(whoami) BRANCH=v1.20.0 ./scripts/travis/build-all-in-one-image.sh
```

results:
https://hub.docker.com/u/morlay/all-in-one/tags
https://hub.docker.com/u/morlay/all-in-one-debug/tags
https://hub.docker.com/u/morlay/opentelemetry-all-in-one/tags

s390x included following jaeger-operator did. 

btw, Why changed base image of `all-in-one` from `scratch` to `alpine:3.12` ? 
i reverted to `scratch` in this pr too.

File: CONTRIBUTING.md
Comment (jpkrohling) [inline CONTRIBUTING.md:179]: nit: can you keep the title without a link, and add the link to "OCI images" in the line 181?

File: CONTRIBUTING.md
Comment (jpkrohling) [inline CONTRIBUTING.md:184]: This isn't the operator repository ;-) The instructions here need to fit this repo.

File: plugin/storage/es/mappings/gen_assets.go
Comment (jpkrohling) [inline plugin/storage/es/mappings/gen_assets.go:216]: This change does not belong to this PR.

File: scripts/travis/build-all-in-one-image.sh
Comment (jpkrohling) [inline scripts/travis/build-all-in-one-image.sh:66]: @aebirim, for your change, you might just need to add a new tag here, as well as a `docker login` to all configured repositories.

File: scripts/travis/build-all-in-one-image.sh
Comment (jpkrohling) [inline scripts/travis/build-all-in-one-image.sh:96]: What's the difference in the clock time for this step with all arches vs. only one arch?

File: scripts/travis/build-all-in-one-image.sh
Comment (jpkrohling) [inline scripts/travis/build-all-in-one-image.sh:103]: I see potential for a refactoring here, which we should do in a next PR.

File: scripts/travis/build-all-in-one-image.sh
Comment (morlay) [reply scripts/travis/build-all-in-one-image.sh:96]: we need to build each go binary for each arch.

the final Dockerfile not includes build stage, just copy binary from the step.

using `for in` here, only want to make ARCHS be easier to configurate

File: scripts/travis/build-all-in-one-image.sh
Comment (morlay) [reply scripts/travis/build-all-in-one-image.sh:103]: Yes. if other components need to support multi-arch. need to mv to a common shell file.

File: scripts/travis/build-all-in-one-image.sh
Comment (jpkrohling) [reply scripts/travis/build-all-in-one-image.sh:96]: My question is mainly to determine if we have a penalty in build time for CI. If it takes 5m to build *each* arch, then we'll have to rethink the approach. If it takes 5m for the first plus 1m for each subsequent arch, then it's fine.

Comment (morlay) [reply scripts/travis/build-all-in-one-image.sh:96]: it will not take 5m for each arch. 

1. go build for amd64, go download deps (3m) + go prebuild common (0.5m) + go prebuild for amd64 (0.5m) + go build bin (1m)
2. go build for arm64, go prebuild for arm64 (0.5m) + go build bin (1m)
2. go build for s390x, go prebuild for s390x (0.5m) + go build bin (1m)

if we split, it will take 5m for each arch. this why i upgrade for jaeger-operator.

Comment (jpkrohling) [reply scripts/travis/build-all-in-one-image.sh:96]: Great, thanks for the confirmation. I wasn't sure whether the download step would be shared across steps. If it adds 90s for each subsequent arch, I think it's reasonable.

File: CONTRIBUTING.md
Comment (jpkrohling) [inline CONTRIBUTING.md:179]: This section needs some better formatting and grammar check (@objectiser?), but I'm fine in having this done by someone else or in a future PR.

**Review by jpkrohling — APPROVED** [review]
I'll leave it open for the day, to give other maintainers the chance to review.

File: cmd/all-in-one/Dockerfile
Comment (Ashmita152) [inline cmd/all-in-one/Dockerfile:8]: Are we changing back the base image back to scratch ?

Comment (jpkrohling) [reply cmd/all-in-one/Dockerfile:8]: Good catch! No, we shouldn't change it back to scratch.

**Review by jpkrohling — COMMENTED** [review]
Dismissing the approval, as the base image shouldn't be changed.

File: cmd/all-in-one/Dockerfile
Comment (morlay) [reply cmd/all-in-one/Dockerfile:8]: i asked. 

Why we need to change to alpine for the release image

Comment (jpkrohling) [reply cmd/all-in-one/Dockerfile:8]: See #2545 for context.

Comment (morlay) [reply cmd/all-in-one/Dockerfile]: Sure. 
got info from https://github.com/jaegertracing/jaeger/issues/2116. 
I have put it back.

File: cmd/all-in-one/Dockerfile
Comment (Ashmita152) [inline cmd/all-in-one/Dockerfile:3]: I see you're using golang:1.15-alpine as base image for debug, but it will still have delve missing. I feel like we reverted back changes to this Dockerfile from my PR.

Comment (jpkrohling) [reply cmd/all-in-one/Dockerfile:3]: This PR here should not revert that PR.

File: cmd/all-in-one/Dockerfile
Comment (Ashmita152) [inline cmd/all-in-one/Dockerfile:57]: I am zero knowledge of docker buildx, curious what is the meaning of have FROM at the end of Dockerfile ?

Comment (morlay) [reply cmd/all-in-one/Dockerfile:57]: Yes. 

it is not new of buildx.

FROM could be from a stage name.

i learn this from istio Dockerfiles, which is used to switch default and distroless

File: cmd/all-in-one/Dockerfile
Comment (morlay) [reply cmd/all-in-one/Dockerfile:11]: Sure. i added it.

We could setup multi arch build for base images next

File: scripts/travis/build-all-in-one-image.sh
Comment (aebirim) [reply scripts/travis/build-all-in-one-image.sh:66]: > @aebirim, for your change, you might just need to add a new tag here, as well as a `docker login` to all configured repositories.

np. @albertteoh @jpkrohling  I'll wait for this script to be merged before adding my quay change

File: CONTRIBUTING.md
Comment (jpkrohling) [inline CONTRIBUTING.md:185]: Why is this `TRAVIS_SECURE_ENV_VARS=true` needed? Isn't there a make target that could be used instead of the Travis script?

**Review by jpkrohling — COMMENTED** [review]
@Ashmita152, could you please review and test this as well? We should have no regressions when compared to your previous work.

**Comment by Ashmita152** [comment]
Sure I will give it a try today.

File: scripts/travis/build-all-in-one-image.sh
Comment (Ashmita152) [inline scripts/travis/build-all-in-one-image.sh:59]: @morlay I noticed all occurrences of TARGETARCH is removed from this script but I don't see the place it is passed  as argument to Dockerfile from this script. TARGETARCH is being used in all-in-one's Dockerfile to decide which binary to copy inside container. Is this some arg which buildx populates internally by default ?

File: scripts/travis/build-all-in-one-image.sh
Comment (Ashmita152) [inline scripts/travis/build-all-in-one-image.sh:12]: Does travis support building images for `s390x` architecture ? Doesn't that means it will require base image of this architecture for alpine too ?

File: scripts/travis/setup-docker-buildx.sh
Comment (Ashmita152) [inline scripts/travis/setup-docker-buildx.sh:52]: I noticed you're using `tonistiigi/binfmt` image. I guess this is because it has images for all architecture which we need. Can we use append the tag too `tonistiigi/binfmt:<tag>` Right now it is pulling latest tag and the owner updating the latest image may break Jaeger travis CI.

**Comment by Ashmita152** [comment]
Hi @morlay looks great overall. I've left some minor comments. Would it be possible for you to share the travis run link where you tested your branch.

**Review by Ashmita152 — CHANGES_REQUESTED** [review]

File: scripts/travis/build-all-in-one-image.sh
Comment (morlay) [reply scripts/travis/build-all-in-one-image.sh:59]: TARGETARCH and other platform args will be passed by `buildx`

https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope

**Comment by morlay** [comment]
@Ashmita152  i tested on local by (need to setup buildx)

```shell
TRAVIS_SECURE_ENV_VARS=true NAMESPACE=$(whoami) BRANCH=v1.20.0 ./scripts/travis/build-all-in-one-image.sh
```

File: scripts/travis/build-all-in-one-image.sh
Comment (morlay) [reply scripts/travis/build-all-in-one-image.sh:12]: Yes. We could build base images by `buildx` in next step.

golang and alpine both support `s390x` too. so the base images of jaeger could support `s390x`

the magic is `tonistiigi/binfmt` who installs `qemu`

File: scripts/travis/setup-docker-buildx.sh
Comment (morlay) [reply scripts/travis/setup-docker-buildx.sh:52]: `tonistiigi/binfmt` is not provide the images we need for architecture.

it just install `qemu` as simulator for target arch. 
so we could run arm64 binary on amd64 host. (but it is a very slow way, and we don't use this for compiling)

the images `alpine:3.12`, you could see the arch list in docker hub.

image

This means docker hub store each image for each arch.

When we use `FROM alpine:3.12`, it will pull the matched arch image to local. 

like docker with `--platform` did.
```
docker pull --platform=linux/arm64 alpine:3.12
docker pull --platform=linux/amd64 alpine:3.12
docker pull --platform=linux/s390x alpine:3.12
```

With `buildx`,  the  `FROM alpine:3.12` in Dockerfile, pulled for each arch image too.
And in `RUN apt add git`, `apt` (arm64 binary) need to execute on matched arch (arm64) host.
This why we need `qemu`. with `qemu` we could run `apt` (arm64 binary) on arm64 host. 

And we don't need the tag.  it is a official tool in [buildx doc](https://github.com/docker/buildx#building-multi-platform-images). when this tool is stable, i think it will merge back to docker's namespace.

Comment (Ashmita152) [reply scripts/travis/setup-docker-buildx.sh:52]: I mean you are using `tonistiigi/binfmt` instead of official image `linuxkit/binfmt` Is there any particular reason for that ?

Comment (morlay) [reply scripts/travis/setup-docker-buildx.sh:52]: no special reason, just it is on README of buildx Which is widely used by github actions.

https://github.com/docker/buildx#building-multi-platform-images
image

Comment (morlay) [reply scripts/travis/setup-docker-buildx.sh:52]: May be the version of qemu.

`tonistiigi/binfmt` with qemu@5.0
`linuxkit/binfmt` with qemu@4.2

Comment (Ashmita152) [reply scripts/travis/setup-docker-buildx.sh:52]: Ahh I see, thank you @morlay that answers my question.

File: CONTRIBUTING.md
Comment (morlay) [reply CONTRIBUTING.md:185]: I don't want to change too much in this pr.

We could refactor this in future

**Comment by Ashmita152** [comment]
I tested the all-in-one's Dockerfile changes locally, it worked for me like charm. Thank you @morlay for answering all my questions. I am not sure what will be the right way to test travis part.

**Comment by morlay** [comment]
`./scripts/travis/build-all-in-one-image.sh` called in pr checks, only difference is not push to hub.

**Comment by jpkrohling** [comment]
@yurishkuro, would you like to take a look? The only point that I had was about the usage of travis-specific env vars for local building/publishing, but if @morlay promises to work on that on a follow-up PR, I'm OK with that.

**Comment by jpkrohling** [comment]
ping @yurishkuro

**Comment by yurishkuro** [comment]
It seems to be that there are a lot of changes to the Dockerfiles that revert the previous changes for debug images.

**Comment by morlay** [comment]
@yurishkuro it is not revert. 

Just added stages of base images  to support multi arch for `all-in-one` images, and enhance the build stages to avoid redundant configuration.

We could to do next for support multi arch for all component images.

* [ ] build & push base & debug muti images 
* [ ] refactor Dockerfile of all-in-one and other components.

This pr is since July. before debug images created.
I don't want change the scope for `all-in-one`.

If you couldn't like those changes. We could close this pr.
And re-open a new pr to setup from base images.

**Comment by yurishkuro** [comment]
It most definitely reverts a bunch of stuff. I realize it is unfortunate that the two PRs overlapped in time, but the debug images PR did a lot of refactoring on the images to DRY them, and those changes are being reverted here (for example, all the copying of cert files, which is currently done only once in the base images).

I think we need to rethink how buildx should work with the new refactored setup, simply merging this old PR on top is probably not going to work.

**Comment by Ashmita152** [comment]
I was thinking to pick work on multi-arch images for base images after this PR gets merged. I can pick it before if you think that will help.

**Comment by morlay** [comment]
@Ashmita152 @yurishkuro 

if we want to setup from base images, and for all components images

Stages could be

Stage 0 - switch docker build to docker buildx build

- updates all docker cli 
- --platform
- Dockerfile，remove default value of platform args like TARGETARCH
- Makefile enhance and updates CI scripts

Base images not need to
publish docker.io.
Using local registery works well too.
But will be a extra setup step for local buld.

Stage 1 - Dockerfile enhance

- multi stage for base image switching

File: scripts/travis/build-all-in-one-image.sh
Comment (jpkrohling) [reply scripts/travis/build-all-in-one-image.sh:66]: This PR was closed in favor of #2611

**Pull request closed without merging**
